synapse/rest/client/v2_alpha/report_event.py

```diff
@@ -58,17 +81,108 @@ async def on_POST(self, request, room_id, event_id):
         )
 
         return 200, {}
 
+    async def _target_room_moderators(self, room_id, event_id, user_id, reason, score):
+        # We're dispatching the abuse report to room moderators.
+        # First, let's make sure that we should.
+
+        # A little sanity check on the event itself.
+        event = await self.store.get_event(event_id)
+        if event.room_id != room_id:
+            raise SynapseError(
+                HTTPStatus.BAD_REQUEST,
+                "No such event in this room",
+                Codes.NOT_FOUND
+            )
+
+        # Now make sure that the user was able to witness the event.
+        events = await filter_events_for_client(
+            self.storage,
+            user_id,
+            [event],
+            is_peeking=True
+        )
+        if len(events) == 0:
+            raise SynapseError(
+                HTTPStatus.FORBIDDEN,
+                "User cannot witness this event",
+                Codes.FORBIDDEN
+            )
+
+        # Find the moderators
+        power_level_event = await self.state.get_current_state(
+            room_id, EventTypes.PowerLevels, ""
+        )
+        ban_level = power_level_event.content.get("ban", 50)
+        kick_level = power_level_event.content.get("kick", 50)
+        moderator_level = max(ban_level, kick_level)
+
+        # map: homeserver:str -> [user_id]
+        room = await self.store.get_room_with_stats(room_id)
+        # FIXME: What do we do if `room is None`?
+
+        moderators_by_hs = {}
+        for member_id in room["joined_members"]:
+            level = power_level_event.content.get("users", {}).get(member_id)
+            if not level:
+                level = power_level_event.content.get("users_default", 0)
+            if level >= moderator_level:
+                hs = UserID.from_string(user_id).domain
+                moderators_on_hs = moderators_by_hs.get(hs, [])
+                moderators_on_hs.append(member_id)
+
+
+        # Dispatch report
+        for (hs, moderators) in moderators_by_hs.items():
+            if hs == self.hs.domain:
+                # Dispatch report immediately as a server notice.
+                for member_id in moderators:
+                    await self.snm.send_notice(
+                        user_id=member_id,
+                        type=EventTypes.Message,
+                        event_content= {
+                            "body": "User has reported content",
+                            "msgtype": "m.server_notice.content_report",
+                            "roomId": room_id,
+                            "eventId": event_id,
+                            "userId": user_id,
+                            "score": score,
+                            "reason": reason
+                        }
+                    )
+            else:
+                # Dispatch report to remote homeserver.
+                # FIXME: Dispatch report to remote moderators.
+                pass
```

Comment:
You'll want to do something with the federation handler here most likely. Reading the MSC I don't know if there's quite enough information in there to implement this.